### ISSUE TITLE
[PATCH] 4497a11 Export DynaLoader symbols from libperl again

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11674]
Migrated from [rt.perl.org#100138](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=100138) (status was 'resolved')

Searchable as RT100138$

**Comment by p5pRT** [comment]
### From @rurban

This is a bug report for perl from rurban@&#8203;cpan\.org\,
generated with the help of perlbug 1\.39 running under perl 5\.15\.3\.

With 5\.15\.2 and the new xubpp\, DynaLoader symbols were XS\_INTERNAL\,
before they were effectively XS\_EXTERNAL\. This broke B&#8203;::C and possibly
other embedded apps which link to DynaLoader functions\.



---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=critical
&nbsp;    module=DynaLoader


---
Site configuration information for perl 5\.15\.3&#8203;:

Configured by rurban at Wed Sep 21 14&#8203;:46&#8203;:58 CEST 2011\.

Summary of my perl5 \(revision 5 version 15 subversion 3\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 818a19bd7b67b794b43e09c66cea27902850c71b
&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=10\.8\.0\, archname=darwin\-nothread\-git
&nbsp;    uname='darwin air\.local 10\.8\.0 darwin kernel version 10\.8\.0&#8203;: tue
jun 7 16&#8203;:33&#8203;:36 pdt 2011; root&#8203;:xnu\-1504\.15\.3~1release\_i386 i386 '
&nbsp;    config\_args='\-de \-Dusedtrace \-Dmksymlinks \-Accflags=\-m64
\-Aldflags=\-m64 \-Dcc=gcc\-mp\-4\.6 \-Dld=gcc\-mp\-4\.6 \-Dusedevel
\-Duseshrplib'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc\-mp\-4\.6'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-m64
\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include
\-I/opt/local/include'\,
&nbsp;    optimize='\-O3'\,
&nbsp;    cppflags='\-fno\-common \-DPERL\_DARWIN \-m64 \-fno\-strict\-aliasing
\-pipe \-fstack\-protector \-I/usr/local/include \-I/opt/local/include'
&nbsp;    ccversion=''\, gccversion='4\.6\.1 20110325 \(prerelease\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc\-mp\-4\.6'\, ldflags =' \-m64 \-fstack\-protector
\-L/usr/local/lib \-L/opt/local/lib'
&nbsp;    libpth=/usr/local/lib /opt/local/lib /usr/lib
&nbsp;    libs=\-lgdbm \-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;    perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;    libc=\, so=dylib\, useshrplib=true\, libperl=libperl\.dylib
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup
\-L/usr/local/lib \-L/opt/local/lib \-fstack\-protector'

Locally applied patches&#8203;:




---
@&#8203;INC for perl 5\.15\.3&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.3/darwin\-nothread\-git
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.3
&nbsp;    /usr/local/lib/perl5/5\.15\.3/darwin\-nothread\-git
&nbsp;    /usr/local/lib/perl5/5\.15\.3
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl 5\.15\.3&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;    HOME=/Users/rurban
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_CTYPE=en\_US\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/opt/local/bin&#8203;:/opt/local/sbin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/Users/rurban/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @rurban

<details><summary>0001-Export-DynaLoader-symbols-from-libperl-again.patch</summary>

```diff
From 4497a11a0d2e48af795e0e4d8f8c185e48d87b10 Mon Sep 17 00:00:00 2001
From: Reini Urban <rurban@x-ray.at>
Date: Tue, 27 Sep 2011 06:45:37 -0500
Subject: [PATCH] Export DynaLoader symbols from libperl again
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------1.7.6.1"

This is a multi-part message in MIME format.
--------------1.7.6.1
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


With 5.15.2 and the new xubpp, DynaLoader symbols were XS_INTERNAL, before
they were effectively XS_EXTERNAL. This broke B::C and possibly other
embedded apps which link to DynaLoader functions.
---
 ext/DynaLoader/dlutils.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)


--------------1.7.6.1
Content-Type: text/x-patch; name="0001-Export-DynaLoader-symbols-from-libperl-again.patch"
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="0001-Export-DynaLoader-symbols-from-libperl-again.patch"

diff --git a/ext/DynaLoader/dlutils.c b/ext/DynaLoader/dlutils.c
index 1ba9a61..574ccad 100644
--- a/ext/DynaLoader/dlutils.c
+++ b/ext/DynaLoader/dlutils.c
@@ -8,6 +8,7 @@
  *                      files when the interpreter exits
  */
 
+#define PERL_EUPXS_ALWAYS_EXPORT
 #ifndef START_MY_CXT /* Some IDEs try compiling this standalone. */
 #   include "EXTERN.h"
 #   include "perl.h"

--------------1.7.6.1--



```
</details>

**Comment by p5pRT** [comment]
### From @Leont

On Tue\, Sep 27\, 2011 at 2&#8203;:52 PM\, Reini Urban \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> With 5\.15\.2 and the new xubpp\, DynaLoader symbols were XS\_INTERNAL\,
> before they were effectively XS\_EXTERNAL\. This broke B&#8203;::C and possibly
> other embedded apps which link to DynaLoader functions\.

Yeah\, if boot\_Dynaloader isn't exported then any embedded perl
application would break\.

Leon

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rurban

On Tue\, Sep 27\, 2011 at 8&#8203;:03 AM\, Leon Timmermans \<fawaka@&#8203;gmail\.com> wrote&#8203;:
> On Tue\, Sep 27\, 2011 at 2&#8203;:52 PM\, Reini Urban \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> With 5\.15\.2 and the new xubpp\, DynaLoader symbols were XS\_INTERNAL\,
>> before they were effectively XS\_EXTERNAL\. This broke B&#8203;::C and possibly
>> other embedded apps which link to DynaLoader functions\.
>
> Yeah\, if boot\_Dynaloader isn't exported then any embedded perl
> application would break\.

boot\_DynaLoader is not the problem\, this is always exported\.
I link to the functions directly\, as to the other perl internal XS functions
\(\_XS\_
The individual symbols are&#8203;:

Undefined symbols&#8203;:
B&#8203;::C svn
$ t/testc\.sh 16
/usr/local/bin/perl5\.15\.2d\-nt \-Mblib script/cc\_harness \-g3 ccode16\.c \-o ccode16
gcc\-mp\-4\.6 \-fno\-common \-DPERL\_DARWIN \-m64 \-DDEBUGGING
\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include
\-I/opt/local/include \-m64
\-I/usr/local/lib/perl5/5\.15\.2/darwin\-debug/CORE \-g3 ccode16\.c \-o
ccode16    \-m64  \-L/usr/local/lib \-L/opt/local/lib \-m64
\-L/usr/local/lib/perl5/5\.15\.2/darwin\-debug/CORE \-lperl \-ldl \-lm \-lutil
\-lc
ccode16\.c&#8203;: In function 'perl\_init\_aaaa'&#8203;:
ccode16\.c&#8203;:5013&#8203;:12&#8203;: note&#8203;: variable tracking size limit exceeded with
\-fvar\-tracking\-assignments\, retrying without
Undefined symbols&#8203;:
&nbsp;  "\_XS\_DynaLoader\_dl\_error"\, referenced from&#8203;:
&nbsp;      \_perl\_init\_aaaa in cc4LCSbS\.o
&nbsp;  "\_XS\_DynaLoader\_dl\_install\_xsub"\, referenced from&#8203;:
&nbsp;      \_perl\_init\_aaaa in cc4LCSbS\.o
&nbsp;  "\_XS\_DynaLoader\_dl\_find\_symbol"\, referenced from&#8203;:
&nbsp;      \_perl\_init\_aaaa in cc4LCSbS\.o
&nbsp;  "\_XS\_DynaLoader\_dl\_undef\_symbols"\, referenced from&#8203;:
&nbsp;      \_perl\_init\_aaaa in cc4LCSbS\.o
&nbsp;  "\_XS\_DynaLoader\_dl\_load\_file"\, referenced from&#8203;:
&nbsp;      \_perl\_init\_aaaa in cc4LCSbS\.o
ld&#8203;: symbol\(s\) not found

$ ls /usr/local/lib/perl5/5\.15\*/darwin\-\*/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.0/darwin\-multi\-debug/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.1/darwin\-debug\-methopt/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.1/darwin\-debug\-multi/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.1/darwin\-debug/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.2/darwin\-debug/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.2/darwin\-nothread\-git/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.3/darwin\-debug/CORE/libperl\.dylib
/usr/local/lib/perl5/5\.15\.3/darwin\-nothread\-git/CORE/libperl\.dylib

$ nm /usr/local/lib/perl5/5\.15\*/darwin\-\*/CORE/libperl\.dylib |grep
XS\_DynaLoader\_dl\_error
0000000000377b9a T \_XS\_DynaLoader\_dl\_error
000000000034b340 T \_XS\_DynaLoader\_dl\_error
00000000001ddb10 T \_XS\_DynaLoader\_dl\_error
000000000034a440 T \_XS\_DynaLoader\_dl\_error
000000000034ee60 t \_XS\_DynaLoader\_dl\_error
000000000012ea80 t \_XS\_DynaLoader\_dl\_error
000000000034bc84 t \_XS\_DynaLoader\_dl\_error
000000000012d3d0 t \_XS\_DynaLoader\_dl\_error

\< 5\.15\.2 T\, since 5\.15\.2 t

DynaLoader is the only XS here with xsubpp mangling\,
the others are proper \.c files\.

$ nm /usr/local/lib/perl5/5\.15\.3/darwin\-debug/CORE/libperl\.dylib |grep \_XS\_

000000000034ee60 t \_XS\_DynaLoader\_dl\_error
000000000034e2e4 t \_XS\_DynaLoader\_dl\_find\_symbol
000000000034e845 t \_XS\_DynaLoader\_dl\_install\_xsub
000000000034dadd t \_XS\_DynaLoader\_dl\_load\_file
000000000034e788 t \_XS\_DynaLoader\_dl\_undef\_symbols
000000000034dfb4 t \_XS\_DynaLoader\_dl\_unload\_file
0000000000319869 T \_XS\_Internals\_HvREHASH
0000000000316c0d T \_XS\_Internals\_SvREADONLY
0000000000317492 T \_XS\_Internals\_SvREFCNT
00000000003196f7 T \_XS\_Internals\_hash\_seed
000000000031791f T \_XS\_Internals\_hv\_clear\_placehold
00000000003197b0 T \_XS\_Internals\_rehash\_seed
000000000031cd1f T \_XS\_PerlIO\_\_Layer\_\_NoWarnings
000000000031ce53 T \_XS\_PerlIO\_\_Layer\_\_find
0000000000317b67 T \_XS\_PerlIO\_get\_layers
000000000030f286 T \_XS\_UNIVERSAL\_DOES
000000000030f470 T \_XS\_UNIVERSAL\_VERSION
000000000030e89c T \_XS\_UNIVERSAL\_can
000000000030e231 T \_XS\_UNIVERSAL\_isa
000000000015246d T \_XS\_mro\_method\_changed\_in
0000000000319b42 T \_XS\_re\_is\_regexp
000000000031acdc T \_XS\_re\_regexp\_pattern
0000000000319ea8 T \_XS\_re\_regname
000000000031a471 T \_XS\_re\_regnames
0000000000319c9e T \_XS\_re\_regnames\_count
0000000000316201 T \_XS\_utf8\_decode
0000000000316500 T \_XS\_utf8\_downgrade
0000000000316122 T \_XS\_utf8\_encode
0000000000315cb1 T \_XS\_utf8\_is\_utf8
000000000031678f T \_XS\_utf8\_native\_to\_unicode
00000000003169ce T \_XS\_utf8\_unicode\_to\_native
0000000000316374 T \_XS\_utf8\_upgrade
0000000000315e30 T \_XS\_utf8\_valid
00000000003131a2 T \_XS\_version\_boolean
0000000000313553 T \_XS\_version\_is\_alpha
0000000000315977 T \_XS\_version\_is\_qv
000000000031026f T \_XS\_version\_new
0000000000313457 T \_XS\_version\_noop
0000000000312820 T \_XS\_version\_normal
00000000003125c0 T \_XS\_version\_numify
000000000031388d T \_XS\_version\_qv
0000000000312360 T \_XS\_version\_stringify
0000000000312a80 T \_XS\_version\_vcmp

\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @tsee

On 09/27/2011 03&#8203;:40 PM\, Reini Urban wrote&#8203;:
> On Tue\, Sep 27\, 2011 at 8&#8203;:03 AM\, Leon Timmermans\<fawaka@&#8203;gmail\.com>  wrote&#8203;:
>> On Tue\, Sep 27\, 2011 at 2&#8203;:52 PM\, Reini Urban\<perlbug\-followup@&#8203;perl\.org>  wrote&#8203;:
>>> With 5\.15\.2 and the new xubpp\, DynaLoader symbols were XS\_INTERNAL\,
>>> before they were effectively XS\_EXTERNAL\. This broke B&#8203;::C and possibly
>>> other embedded apps which link to DynaLoader functions\.
>>
>> Yeah\, if boot\_Dynaloader isn't exported then any embedded perl
>> application would break\.
>
> boot\_DynaLoader is not the problem\, this is always exported\.
> I link to the functions directly\, as to the other perl internal XS functions
> \(\_XS\_
> The individual symbols are&#8203;:

\[\.\.\.\]

> Undefined symbols&#8203;:
>    "\_XS\_DynaLoader\_dl\_error"\, referenced from&#8203;:
>        \_perl\_init\_aaaa in cc4LCSbS\.o

Makes me wonder whether the exact name of XSUBs is considered an 
implementation detail or part of the API\.

Anyway\, on a recentish blead perl at 
a62da8e6727f34e800a1b8c33214cc22c872ad0e \(Aug 28th\)&#8203;:

tsee@&#8203;l3tsee&#8203;:\~/perl/cpan\_perl\_branch\_smoke$ nm 
\.\./\.\./bleadperl/perl\-ssh/perl | grep XS | grep Dyna
0000000000710dc2 t XS\_DynaLoader\_CLONE
0000000000710b5a t XS\_DynaLoader\_dl\_error
000000000071004a t XS\_DynaLoader\_dl\_find\_symbol
00000000007105b7 t XS\_DynaLoader\_dl\_install\_xsub
000000000070f7e9 t XS\_DynaLoader\_dl\_load\_file
00000000007104fd t XS\_DynaLoader\_dl\_undef\_symbols
000000000070fcf8 t XS\_DynaLoader\_dl\_unload\_file

In a nutshell\, XSUBs should now be exported alright if

a\) xsubpp is old\.
b\) They aren't generated by xsubpp \(duh\)\, even if they're declared with 
the XS\(\) macro\.
c\) They are generated by xsubpp\, but the XS code
uses the "EXPORT\_XSUB\_SYMBOLS&#8203;: ENABLE" XS verb to disable generation of 
'static' XSUBs\.
d\) \(And if this breaks\, you get to keep both pieces\) you do "\#define 
PERL\_EUPXS\_ALWAYS\_EXPORT" so that it's defined during the XSUB 
declaration and implementation\. It basically forces the ParseXS\-internal 
variant of the XS\(\) macro to use XS\_EXTERNAL instead of XS\_INTERNAL\.

Does that give you enough rope to hang anything?

Cheers\,
Steffen

**Comment by p5pRT** [comment]
### From @rurban

On Wed\, Sep 28\, 2011 at 1&#8203;:31 AM\, Steffen Mueller
\<smueller@&#8203;ma2geo\.mathematik\.uni\-karlsruhe\.de> wrote&#8203;:
> d\) \(And if this breaks\, you get to keep both pieces\) you do "\#define
> PERL\_EUPXS\_ALWAYS\_EXPORT" so that it's defined during the XSUB declaration
> and implementation\. It basically forces the ParseXS\-internal variant of the
> XS\(\) macro to use XS\_EXTERNAL instead of XS\_INTERNAL\.
>
> Does that give you enough rope to hang anything?

Didn't you see my patch in the ticket? 5\.15\.2 and 5\.15\.3 is already
broken\. Please let not 5\.15\.4 also be broken\.

I added \#define PERL\_EUPXS\_ALWAYS\_EXPORT to dlutils\.c so all
DynaLoader\.xs versions export now their symbols as before\, and I can
link to them as before\.

The problem is in CORE\, all other things I can manage by myself\.
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @rurban

Can someone please apply the patch?
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Sep 27 05&#8203;:52&#8203;:30 2011\, rurban wrote&#8203;:
> This is a bug report for perl from rurban@&#8203;cpan\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.15\.3\.
> 
> With 5\.15\.2 and the new xubpp\, DynaLoader symbols were XS\_INTERNAL\,
> before they were effectively XS\_EXTERNAL\. This broke B&#8203;::C and possibly
> other embedded apps which link to DynaLoader functions\.

Thank you\.  I have applied your attached patch as 2e34687\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From kaffeetisch@gmx.de

On 27\.09\.2011 14&#8203;:52\, Reini Urban \(via RT\) wrote&#8203;:
> \+\#define PERL\_EUPXS\_ALWAYS\_EXPORT

If I read Steffen's reply and commit 
a62da8e6727f34e800a1b8c33214cc22c872ad0e correctly\, this change 
shouldn't be necessary anymore\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Sep 30 16&#8203;:27&#8203;:52 2011\, kaffeetisch@&#8203;gmx\.de wrote&#8203;:
> On 27\.09\.2011 14&#8203;:52\, Reini Urban \(via RT\) wrote&#8203;:
> > \+\#define PERL\_EUPXS\_ALWAYS\_EXPORT
> 
> If I read Steffen's reply and commit 
> a62da8e6727f34e800a1b8c33214cc22c872ad0e correctly\, this change 
> shouldn't be necessary anymore\.
> 

Does that mean I should revert Reini’s patch which I applied a few hours
ago?

**Comment by p5pRT** [comment]
### From @rurban

On Fri Sep 30 16&#8203;:27&#8203;:52 2011\, kaffeetisch@&#8203;gmx\.de wrote&#8203;:
> On 27\.09\.2011 14&#8203;:52\, Reini Urban \(via RT\) wrote&#8203;:
> > \+\#define PERL\_EUPXS\_ALWAYS\_EXPORT
>
> If I read Steffen's reply and commit
> a62da8e6727f34e800a1b8c33214cc22c872ad0e correctly\, this change
> shouldn't be necessary anymore\.

Wrong\. This change caused the DynaLoader breakage\.

On Fri\, Sep 30\, 2011 at 8&#8203;:10 PM\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>Does that mean I should revert Reini’s patch which I applied a few hours
ago?

No\. Thanks for applying\.

BTW&#8203;: I've just setup a B&#8203;::C smoker here at cPanel which tests blead
daily\, so I hopefully catch breakage earlier\.
I'm not sure how to add CORE tests to such weird cases\, but I'm thinking of it\.
It will help me most :\)
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Sep 30 18&#8203;:48&#8203;:42 2011\, rurban wrote&#8203;:
> BTW&#8203;: I've just setup a B&#8203;::C smoker here at cPanel which tests blead
> daily\, so I hopefully catch breakage earlier\.
> I'm not sure how to add CORE tests to such weird cases\, but I'm
> thinking of it\.
> It will help me most :\)

Thank you\.  That sort of thing is always helpful\.

\(I test my CPAN modules monthly with blead\.  I’ve caught quite a few
bugs that way\. :\-\)

**Comment by p5pRT** [comment]
### From kaffeetisch@gmx.de

On 01\.10\.2011 03&#8203;:48\, Reini Urban wrote&#8203;:
> On Fri Sep 30 16&#8203;:27&#8203;:52 2011\, kaffeetisch@&#8203;gmx\.de wrote&#8203;:
>> On 27\.09\.2011 14&#8203;:52\, Reini Urban \(via RT\) wrote&#8203;:
>>> \+\#define PERL\_EUPXS\_ALWAYS\_EXPORT
>>
>> If I read Steffen's reply and commit
>> a62da8e6727f34e800a1b8c33214cc22c872ad0e correctly\, this change
>> shouldn't be necessary anymore\.
>
> Wrong\. This change caused the DynaLoader breakage\.

While it wasn't actually this commit that changed the visibility of 
DynaLoader symbols\, you're right that it doesn't fix it either\.  \(Since 
DynaLoader goes through xsubpp normally\, so it doesn't use the XS macro 
manually which is the only case affected by the cited commit and its 
parent\.\)

**Comment by p5pRT** [comment]
### From @tsee

Hi all\,

this is a mixed response to Reini and Thorsten\, sorry about that\.

On 10/01/2011 11&#8203;:38 AM\, Torsten Schoenfeld wrote&#8203;:
> On 01\.10\.2011 03&#8203;:48\, Reini Urban wrote&#8203;:
>> On Fri Sep 30 16&#8203;:27&#8203;:52 2011\, kaffeetisch@&#8203;gmx\.de wrote&#8203;:
>>> On 27\.09\.2011 14&#8203;:52\, Reini Urban \(via RT\) wrote&#8203;:
>>>> \+\#define PERL\_EUPXS\_ALWAYS\_EXPORT

Sorry for being slow in applying\. I had planned to think about a more 
elegant solution and figuring out how DynaLoader / B&#8203;::C are special 
before applying this\. Evidently\, I did not get to that in time\.

>>> If I read Steffen's reply and commit
>>> a62da8e6727f34e800a1b8c33214cc22c872ad0e correctly\, this change
>>> shouldn't be necessary anymore\.
>>
>> Wrong\. This change caused the DynaLoader breakage\.

Though is it actually DynaLoader breakage or B&#8203;::C breakage? Either way\, 
adding the define seems like a horrible but acceptable workaround that 
is less work than adding the explicit XS verb to each of the various \.xs 
files of DynaLoader\. Including a comment as to what the define does 
wouldn't have hurt\, though\.

> While it wasn't actually this commit that changed the visibility of
> DynaLoader symbols\, you're right that it doesn't fix it either\. \(Since
> DynaLoader goes through xsubpp normally\, so it doesn't use the XS macro
> manually which is the only case affected by the cited commit and its
> parent\.\)

Indeed\. It seemed fishy to me that the define in a \.c file would do 
anything at all\. But since that's \#included by the various \.xs files\, it 
does work\.

\-\-Steffen

**Comment by p5pRT** [comment]
### From @rurban

On Sat\, Oct 1\, 2011 at 7&#8203;:23 AM\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
> Hi all\,
>
> this is a mixed response to Reini and Thorsten\, sorry about that\.
>
> On 10/01/2011 11&#8203;:38 AM\, Torsten Schoenfeld wrote&#8203;:
>>
>> On 01\.10\.2011 03&#8203;:48\, Reini Urban wrote&#8203;:
>>>
>>> On Fri Sep 30 16&#8203;:27&#8203;:52 2011\, kaffeetisch@&#8203;gmx\.de wrote&#8203;:
>>>>
>>>> On 27\.09\.2011 14&#8203;:52\, Reini Urban \(via RT\) wrote&#8203;:
>>>>>
>>>>> \+\#define PERL\_EUPXS\_ALWAYS\_EXPORT
>
> Sorry for being slow in applying\. I had planned to think about a more
> elegant solution and figuring out how DynaLoader / B&#8203;::C are special before
> applying this\. Evidently\, I did not get to that in time\.
>
>>>> If I read Steffen's reply and commit
>>>> a62da8e6727f34e800a1b8c33214cc22c872ad0e correctly\, this change
>>>> shouldn't be necessary anymore\.
>>>
>>> Wrong\. This change caused the DynaLoader breakage\.
>
> Though is it actually DynaLoader breakage or B&#8203;::C breakage?

DynaLoader works fine\. Only perlembed apps which link to DynaLoader broke\.
The only to me known such app is B&#8203;::C\. But I haven't checked the others on cpan\.
I guess they go through the perl interface\, not the C interface\.

> Either way\,
> adding the define seems like a horrible but acceptable workaround that is
> less work than adding the explicit XS verb to each of the various \.xs files
> of DynaLoader\. Including a comment as to what the define does wouldn't have
> hurt\, though\.

I though the define is explicit enough\. And there is also this ticket
and the commit message\, which is also explicit\.

>> While it wasn't actually this commit that changed the visibility of
>> DynaLoader symbols\, you're right that it doesn't fix it either\. \(Since
>> DynaLoader goes through xsubpp normally\, so it doesn't use the XS macro
>> manually which is the only case affected by the cited commit and its
>> parent\.\)
>
> Indeed\. It seemed fishy to me that the define in a \.c file would do anything
> at all\. But since that's \#included by the various \.xs files\, it does work\.
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/
